Microphone emulation

RipleyTom opened the pull request:
Supports 4 microphone types:
-Standard: your basic microphone
-Singstar: fakes a singstar microphone device
-Real Singstar: if you want to use your real singstar microphones
-Rocksmith: to use with rocksmith games and your rocksmith dongle( thanks @jjsat )

Configuration is under the Audio tab.

RipleyTom commented:
Good catch, added a tooltip.

LuismaSP89 commented:
Great job :P another ps3 accesory emulated. If you want to try next with Buzz controllers you have my blessing xD

Megamouse commented on rpcs3/Emu/Cell/Modules/cellAvconfExt.cpp:
```diff
@@ -38,17 +110,14 @@ s32 cellVideoOutSetupDisplay()
+	std::vector<device_in> ain_devices;
+	get_vec_input_devices(ain_devices);
+
+	if (ain_devices.size() == 0 || deviceNumber > (ain_devices.size() - 1)) return CELL_AUDIO_OUT_ERROR_DEVICE_NOT_FOUND;
```
can you put the return in a new line or scope?

kd-11 commented on rpcs3/Emu/Cell/Modules/cellAvconfExt.cpp:
```diff
@@ -38,17 +110,17 @@ s32 cellVideoOutSetupDisplay()
-	strcpy(info->name, "USB Camera");
+
+	std::vector<device_in> ain_devices;
+	get_vec_input_devices(ain_devices);
```
Why not just `std::vector<> ain_devices = get_input_devices()`? The compiler always avoids the copy as a basic optimization (Its part of the c++ compiler spec, copy ellision)

elad335 commented on rpcs3/Emu/Cell/Modules/cellMic.cpp:
```diff
@@ -8,23 +8,212 @@
+	switch (bytesize)
+	{
+	case 4:
+		dst[0] = src[3];
```
Dont byteswap by hand, src and dst can be void* for easier casting.
```suggestion
		*(u32*)dst = *(be_t<u32>*)src;
```

psennermann commented:
"Singstar vol.1" works really good with any microphone, while "Singstar Vasco" doesn't work (although I have an original Singstar USB converter attached to the usb port, that game keeps on asking to attach one…)

The only minor problems I encountered on my PC with "Singstar vol. 1" are:

- FMV video is out of sync compared to audio (but fortunately text words are perfectly in sync)

- using 2 Singstar microphones with the usb converter they're recognised as one (like what happens in the Windows 10 audio manager); so if you wanna play duett you have to attach another usb microphone...

- when you want to rehear the singed song, the music is all messed up (maybe #5456 could help in this case?)

RipleyTom commented:
I added support(untested) for real Singstar devices.
Also added UI so you can select the microphones you want to use specifically.
As for Singstar Vasco I didn't have time to investigate why it doesn't recognize the device yet.

psennermann commented:
Using "Real Singstar" option, microphone doesn't work well (it has a "muffled/garbled" sound somewhat like what you hear when you want to rehear the singed song), while using two microphones attached to the Singstar usb converter, every microphone outputs voice for both blue and red microphone (like you were singing with both at the same time)

RipleyTom commented:
Immediately realized my error, real singstar issue should be fixed with the version i just pushed.
Thanks for testing it.

psennermann commented:
Yes, fantastic job, now with "Real Singstar" option the blue and the red microphones have separated output like on real hardware

Megamouse commented on rpcs3/rpcs3qt/settings_dialog.cpp:
```diff
@@ -730,6 +730,80 @@ settings_dialog::settings_dialog(std::shared_ptr<gui_settings> guiSettings, std:
+	auto ChangeMicrophoneDevice = [=](u32 next_index, QString text)
+	{
+		xemu_settings->SetSetting(emu_settings::MicrophoneDevices, xemu_settings->m_microphone_creator.SetDevice(next_index, text));
+		if(next_index < 4 && text == xemu_settings->m_microphone_creator.mic_none)
```
```suggestion
		if (next_index < 4 && text == xemu_settings->m_microphone_creator.mic_none)
```

RipleyTom commented:
I looked at Singstar Vasco and my guess is it requires a proper implementation of sys_usbd which we don't have atm and it's outside the scope of this PR so it won't be fixed by this(but I'm looking into it).

RipleyTom commented:
@psennermann If you could get me a dump of the USB descriptor of the singstar usb device with https://www.thesycon.de/eng/usb_descriptordumper.shtml it would be greatly appreciated.
Just paste all the text on something like https://pastebin.com/ and put the link here, tyvm!

psennermann commented:
Here you are:

https://pastebin.com/zUQU2k50

LonelyDragon123 commented:
Getting a fatal error on Guitar Hero: Warriors of Rock and World Tour
![default](https://user-images.githubusercontent.com/28080230/51363820-c6b60d80-1aea-11e9-9d3c-46a3208c3c8e.png)

[RPCS3.log](https://github.com/RPCS3/rpcs3/files/2771518/RPCS3.log)

psennermann commented:
Any news about this PR?

RipleyTom commented:
I've started working on something else(probably PR next week) then I have something related to do in another PR and then I'll come back to this.
Don't worry I haven't forgotten about it(mostly got sidetracked because of singstar vasco and then sidetracked again for another reason).

uaqlover commented:
@RipleyTom thanks for you hard work i appreciate it

hectaacdc commented:
Vocals on both Rock Band [BLUS30050] and Rock Band 2 [BLUS30147], and any other Rock Band Track Pack works perfectly, this is the only instrument missing on this games, I hope this can get merged soon 👍 
![rb1-vocals](https://user-images.githubusercontent.com/49326873/55661559-d0463c00-57e2-11e9-918d-a6dfb53d8221.png)
![rb2-vocals](https://user-images.githubusercontent.com/49326873/55661612-1b604f00-57e3-11e9-8ebc-ba757e49475a.png)

jjsat commented:
I added some code for using a Rocksmith USB cable with this PR. That cable acts as a mono microphone. The code is enough to get the game to recognize an attached guitar and to process input from it. Not sure if it entirely correct though since the game has a lot of audio output stutter (disabling framelimit and using Null renderer helps a bit).

There is one big hack in cellMicOpenEx, which Rocksmith 2014 calls with rawChannel=4. This value is ignored and hard set to 1 for mono input. Not sure what to make of this, maybe it is more like maxChannels in cellMicOpenRaw?

Changes are here: https://github.com/jjsat/rpcs3/commit/70119abc4689270ecc86d6a85aa5da3c0f52fa1b

RipleyTom commented:
Did a bit of spring cleaning, tested on linux, added rocksmith support as suggested by @jjsat .
Biggest surprise is that it seems to now work perfectly with Karaoke Revolution which I couldn't figure the issues with. Apparently the issues were not with my cellMic implementation so I'm removing the [WIP].

psennermann commented:
No problems to report

LonelyDragon123 commented:
I have problem in Guitar Hero. Even with original Logitech guitar hero mic I have loud noise, and if I say something it changes a little, which means mic working, but sound input very distorted.

kd-11 commented on rpcs3/Emu/Cell/Modules/cellAvconfExt.cpp:
```diff
@@ -73,21 +180,23 @@ s32 cellVideoOutGetGamma(u32 videoOut, vm::ptr<f32> gamma)
+	}
+
+	auto av_manager = fxm::get_always<avconf_manager>();
+
```
Doesn't just changing count to be `count = std::min<u32>(count, av_manager->devices.size())` make things a bit easier? Simplifies the for loop comparison as well as the return statement.

RipleyTom replied:
It does indeed, I went the overly complicated way.

kd-11 commented on rpcs3/Emu/Cell/Modules/cellMic.cpp:
```diff
@@ -1,31 +1,398 @@
+	switch (bytesize)
+	{
+	case 4:
+		*(u32*)dst = *(be_t<u32>*)src;
```
You can also just say `dst = se_storage<u32>::swap(src)` btw. This is also fine; just dislike unnecessary indirection.

RipleyTom replied:
I wasn't aware of that function at the time, changing it.

kd-11 commented on rpcs3/Emu/Cell/Modules/cellMic.cpp:
```diff
@@ -1,31 +1,398 @@
+		stop_microphone();
+	}
+
+	for (u32 index = 0; index < input_devices.size(); index++)
```
this loop could have been simplified as `for (auto& device : input_devices)`

kd-11 commented on rpcs3/Emu/Cell/Modules/cellMic.cpp:
```diff
@@ -1,31 +1,398 @@
+
+s32 microphone_device::start_microphone()
+{
+	for (auto micdevice : input_devices)
```
auto&

kd-11 commented on rpcs3/Emu/Cell/Modules/cellMic.cpp:
```diff
@@ -1,31 +1,398 @@
+
+s32 microphone_device::stop_microphone()
+{
+	for (auto micdevice : input_devices)
```
auto&

kd-11 commented on rpcs3/rpcs3qt/settings_dialog.cpp:
```diff
@@ -782,6 +782,82 @@ settings_dialog::settings_dialog(std::shared_ptr<gui_settings> guiSettings, std:
+			return;
+		}
+	};
+	auto PropagateUsedDevices = [=]()
```
Space these lambdas a little. I missed this being a separate function the first time.

kd-11 commented on rpcs3/Emu/Cell/Modules/cellMic.cpp:
```diff
@@ -34,161 +401,350 @@ s32 cellMicInit()
 
-	const auto micThread = fxm::make<mic_thread>("Mic Thread");
-	if (!micThread)
+	auto mic_t = g_idm->lock<mic_thread>(id_new);
```
This abbreviation is used all over the new code and kinda bugs me since the _t suffix usually denotes a type. Why not just mic_thread or even mic_thr? Its a nitpick but its quite triggering to anyone who has worked with C for a long time.

RipleyTom replied:
You're completely right, mic_thr it is!

kd-11 commented on rpcs3/Emu/Cell/Modules/cellMic.cpp:
```diff
@@ -34,161 +401,350 @@ s32 cellMicInit()
+
+		thread_ctrl::wait_for(1000);
+
+		auto mic_thr = g_idm->lock<mic_thread>(0);
```
See? mic_thr looks much better :p
Now if only we had some consistency...

RipleyTom commented:
@LonelyDragon123 make sure you're using standard microphone(in the microphone type combobox in audio tab) with guitar hero. I just tested guitar hero 5 and it worked perfectly.

elad335 commented on rpcs3/Emu/Cell/Modules/cellMic.cpp:
```diff
@@ -232,8 +81,73 @@ void mic_context::operator()()
+{
+	switch (bytesize)
+	{
+	case 4: *(u32*)dst = se_storage<u32>::swap(*((be_t<u32>*)src)); break;
```
It is more correct to use be_t alone (no manual swap), because it obeys both IS_LE_MACHINE and IS_BE_MACHINE.

elad335 replied:
also bugged.

RipleyTom replied:
Fixed thanks!

elad335 commented on rpcs3/Emu/Cell/Modules/cellMic.cpp:
```diff
@@ -467,8 +1008,8 @@ s32 cellMicGetDeviceIdentifier()
 
-DECLARE(ppu_module_manager::cellMic)("cellMic", []()
-{
+DECLARE(ppu_module_manager::cellMic)
```
the old style was OK, this also doesnt have curly brace on new line.

RipleyTom replied:
Pretty sure it's the result of vs code formatting according to our clang format, I'll leave it as is or I'd have to remember to change it back after every formatting.

elad335 commented on rpcs3/Emu/Cell/Modules/cellMic.cpp:
```diff
@@ -384,9 +922,12 @@ s32 cellMicGetDeviceGUID()
-	UNIMPLEMENTED_FUNC(cellMic);
+	cellMic.todo("cellMicGetType(dev_num=%d, ptr_type=*0x%x)", dev_num, ptr_type);
+
+	*ptr_type = CELLMIC_TYPE_USBAUDIO;
```
Check for nullptr?

elad335 commented on rpcs3/Emu/Cell/Modules/cellMic.cpp:
```diff
@@ -234,72 +768,64 @@ s32 cellMicGetSignalState(u32 deviceNumber, CellMicSignalState signalState, vm::
-	auto micQueue = lv2_event_queue::find(key);
-	micQueue->send(0, CELL_MIC_ATTACH, 0, 0);
-	micThread->eventQueueKey = key;
+	auto micQueue            = lv2_event_queue::find(key);
```
Missing nullptr check. (in case key is invalid)
